File: 02/main.py
from collections import namedtuple


Command = namedtuple("Command", "direction, units")


class Location(namedtuple("Location", "distance, depth")):
    def __add__(self, other):
        return Location(self.distance + other.distance, self.depth + other.depth)

    def __mul__(self, number):
        return Location(self.distance * number, self.depth * number)

    def __rmul__(self, number):
        return self * number


def line_to_command(line):
    command, units_string = line.strip().split()
    return Command(command, int(units_string))


def read_puzzle_input(filename):
    with open(filename) as stream:
        return [line_to_command(row) for row in stream.read().strip().split("\n")]


DIRECTION_DICT = {
    "forward": Location(1, 0),
    "up": Location(0, -1),
    "down": Location(0, 1),
}


def follow_course(course, starting_location=Location(0, 0)):
    location = starting_location
    for command in course:
        location += command.units * DIRECTION_DICT[command.direction]
    return location


if __name__ == "__main__":
    puzzle_input = read_puzzle_input("input.txt")
    target_location = follow_course(puzzle_input)
    print(target_location)
    print(target_location.distance * target_location.depth)

File: 02/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Direction int

const (
	forward Direction = iota
	up
	down
	unspecified
)

type Command struct {
	direction Direction
	units     int
}

type Location struct {
	distance int
	depth    int
}

func parseDirection(directionString string) (Direction, error) {
	switch directionString {
	case "up":
		return up, nil
	case "forward":
		return forward, nil
	case "down":
		return down, nil
	}
	return unspecified, fmt.Errorf("The string '%s' was not recognized as a direction.", directionString)
}

func parseCommand(line string) Command {
	s := strings.Fields(line)
	command, err := parseDirection(s[0])
	if err != nil {
		panic(err)
	}
	units, err := strconv.Atoi(s[1])
	if err != nil {
		panic(err)
	}
	return Command{command, units}
}

func readPuzzleInput(filePath string) []Command {
	file, err := os.Open(filePath)
	if err != nil {
		panic(err)
	}
	var commands []Command
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		commands = append(commands, parseCommand(line))
	}
	return commands
}

func followCourse(course []Command) Location {
	location := Location{0, 0}
	for _, command := range course {
		switch command.direction {
		case up:
			location = Location{location.distance, location.depth - command.units}
		case down:
			location = Location{location.distance, location.depth + command.units}
		case forward:
			location = Location{location.distance + command.units, location.depth}
		}
	}
	return location
}

func main() {
	puzzleInput := readPuzzleInput("input.txt")
	targetLocation := followCourse(puzzleInput)
	fmt.Printf("%d\n", targetLocation.depth*targetLocation.distance)
}
